src/schedule/mod.rs

```diff
@@ -1,37 +1,326 @@
 use crate::ciphersuite::*;
 use crate::codec::*;
 use crate::group::*;
+use crate::messages::{proposals::AddProposal, GroupSecrets, PathSecret};
 use crate::tree::index::LeafIndex;
+use crate::tree::index::NodeIndex;
+use crate::tree::private_tree::CommitSecret;
 use crate::tree::secret_tree::SecretTree;
+use crate::tree::treemath;
+use crate::tree::RatchetTree;
 
 use self::errors::KeyScheduleError;
 
 pub mod errors;
 
-#[cfg(test)]
-mod test_schedule;
-
 pub fn derive_secret(ciphersuite: &Ciphersuite, secret: &Secret, label: &str) -> Secret {
     hkdf_expand_label(ciphersuite, secret, label, &[], ciphersuite.hash_length())
 }
 
-pub fn mls_exporter(
-    ciphersuite: &Ciphersuite,
-    epoch_secrets: &EpochSecrets,
-    label: &str,
-    group_context: &GroupContext,
-    key_length: usize,
-) -> Secret {
-    let secret = &epoch_secrets.exporter_secret;
-    let context = &group_context.serialize();
-    let context_hash = &ciphersuite.hash(context);
-    hkdf_expand_label(
-        ciphersuite,
-        &derive_secret(ciphersuite, secret, label),
-        "exporter",
-        context_hash,
-        key_length,
-    )
+/// The `InitSecret` is used to start the connect the next epoch to the current
+/// one. It's necessary to be able clone this to create a provisional group
+/// state, which includes the `InitSecret`.
+#[derive(Debug, Clone)]
+pub struct InitSecret {
+    secret: Secret,
+}
+
+impl InitSecret {
+    /// Derive an `InitSecret` from an `EpochSecret`.
+    fn derive_init_secret(ciphersuite: &Ciphersuite, epoch_secret: &EpochSecret) -> Self {
+        InitSecret {
+            secret: derive_secret(ciphersuite, &epoch_secret.secret, "init"),
+        }
+    }
+
+    /// Sample a fresh, random `InitiSecret` for the creation of a new group.
+    pub(crate) fn from_random(length: usize) -> Self {
+        InitSecret {
+            secret: Secret::from_random(length),
+        }
+    }
+}
+
+/// It's necessary to clone this to be able generate `GroupSecret` object with
+/// multiple different `PathSecret` objects.
+#[derive(Debug, Clone)]
+pub struct JoinerSecret {
+    secret: Secret,
+}
+
+impl JoinerSecret {
+    /// Derive a `JoinerSecret` from a `CommitSecret` and an `EpochSecrets`
+    /// object, which onctains the necessary `InitSecret`. TODO: For now, this
+    /// takes a reference to a `CommitSecret` as input. This should change with
+    /// #224.
+    pub(crate) fn derive_joiner_secret(
+        ciphersuite: &Ciphersuite,
+        commit_secret: &CommitSecret,
+        init_secret: InitSecret,
```

Comment:
Ideally, the secret would always be consumed. There is only one place, where we have to clone it and then call this function, because the group state we're creating is only provisional (you remarked on it in another comment).